File: tdd/scala/scalatest/lib/Money.scala
package lib {

class Bank {
    def apply(money: Money) {
    }
}

class InvalidMoneyOperation extends Exception

abstract class AMoney;
class Money(mValue:Int, mName:Symbol) extends AMoney with Ordered[Money] {
    val value = mValue
    val name = mName

    def +(other: Money) = {
        if(other.name != name) throw new InvalidMoneyOperation()
        new Money(value + other.value, name)
    }

    def -(other: Money) = {
        if(other.name != name) throw new InvalidMoneyOperation()
        new Money(value - other.value, name)
    }

    override def equals(other: Any) = other match {
        case other: Money => (name == other.name && value == other.value)
        case _ => super.equals(other)
    }

    override def compare(other: Money) = if(name == other.name) value.compare(other.value) else -1

    override def <(other: Money) = if(name == other.name) super.<(other) else false
    override def <=(other: Money) = if(name == other.name) super.<=(other) else false

    def apply_on(bank: Bank) {
        bank apply this
    }

    override def toString() = "$: " + value + " " + name
}

}

File: tdd/ruby/rspec/lib/money.rb
class Money
  include Comparable
  attr_accessor :name, :value

  def initialize(value, name)
    @value, @name = value, name
  end

  def +(other)
    raise InvalidMoneyOperation if other.name != name
    return Money.new value + other.value, name
  end

  def -(other)
    raise InvalidMoneyOperation if other.name != name
    return Money.new value - other.value, name
  end

  def <=>(other)
    if name != other.name
      return -1
    end

    value <=> other.value
  end

  def <(other)
    return super if other.name == name
    false
  end

  def <=(other)
    return super if other.name == name
    false
  end

  def apply_on(bank)
    bank.apply(self)
  end

  def inspect
    "$: #@value #@name"
  end
end

class InvalidMoneyOperation < StandardError
end
